File: Easy/202. Happy Number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        table = {}
        while n:
            if n in table: return False
            table[n] = False
            
            num = 0
            while n:
                num += (n % 10)**2
                n //= 10
            n = num
            
            if num == 1: return True
        return False

File: Easy/202. Happy Number.cpp
class Solution {
public:
    bool isHappy(int n) {
        std::unordered_map<int, bool> map;
        while (n) {
            if (map.find(n) == map.end()) {
                map.insert(std::pair<int,bool>(n,true));
            }else{
                return false;
            }

            int num = 0;
            while (n) {
                int forPow = n % 10;
                num += forPow * forPow;
                n /= 10;
            }
            n = num;

            if(num == 1){
                return true;
            }
        }
        return false;
    }
};
